exonum/src/events/noise/mod.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use byteorder::{ByteOrder, LittleEndian};
+use futures::future::{done, Future};
+use tokio_core::net::TcpStream;
+use tokio_io::{AsyncRead, codec::Framed, io::{read_exact, write_all}};
+
+use std::io;
+
+use crypto::{PublicKey, SecretKey};
+use events::codec::MessagesCodec;
+use events::noise::wrapper::{NoiseWrapper, HANDSHAKE_HEADER_LENGTH};
+
+pub mod wrapper;
+
+type HandshakeResult = Box<Future<Item = Framed<TcpStream, MessagesCodec>, Error = io::Error>>;
+
+#[derive(Debug, Clone)]
+/// Params needed to establish secured connection using Noise Protocol.
+pub struct HandshakeParams {
+    pub public_key: PublicKey,
+    pub secret_key: SecretKey,
+    pub max_message_len: u32,
+}
+
+#[derive(Debug)]
+pub struct NoiseHandshake {}
+
+impl NoiseHandshake {
+    pub fn listen(params: &HandshakeParams, stream: TcpStream) -> HandshakeResult {
+        listen_handshake(stream, params)
+    }
+
+    pub fn send(params: &HandshakeParams, stream: TcpStream) -> HandshakeResult {
+        send_handshake(stream, params)
+    }
+}
+
+fn listen_handshake(stream: TcpStream, params: &HandshakeParams) -> HandshakeResult {
+    let max_message_len = params.max_message_len;
+    let mut noise = NoiseWrapper::responder(params);
+    let framed = read(stream).and_then(move |(stream, msg)| {
+        let _buf = noise.read_handshake_msg(&msg);
```

Comment:
Doesn't this (and other similar statements) swallow an error? It would be useful to have at least one test per each step of the handshake, where the counterparty works not according to the Noise protocol. Without such testing, I'd consider the implementation to be hacky at best.

Reply:
Agreed, I added a test in #702.